src/error.rs

```diff
@@ -75,3 +79,106 @@ impl ParseError {
         ParseError::Other(tag.to_string(), reason.to_string())
     }
 }
+
+#[derive(Error)]
+/// Error akin to `ParseError` with context where it occurred.
+pub struct ContextParseError {
+    error: ParseError,
+    context: Option<xmltree::Element>,
+}
+
+impl ContextParseError {
+    /// Create a new error with context from error and context.
+    pub fn new(error: ParseError, context: xmltree::Element) -> Self {
+        Self {
+            error,
+            context: Some(context),
+        }
+    }
+}
+
+impl Debug for ContextParseError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        let context = self
+            .context
+            .as_ref()
+            .map_or(String::from("None"), display_context);
+
+        f.debug_struct("ContextParseError")
+            .field("error", &self.error)
+            .field("context", &context)
+            .finish()
+    }
+}
+
+impl Display for ContextParseError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        write!(f, "{}", self.error)?;
+
+        if let Some(context) = &self.context {
+            write!(f, "\n{}", display_context(context))?;
+        }
+
+        Ok(())
+    }
+}
+
+impl From<ContextParseError> for ParseError {
+    fn from(error: ContextParseError) -> Self {
+        error.error
+    }
+}
+
+impl From<ParseError> for ContextParseError {
+    fn from(error: ParseError) -> Self {
+        Self {
+            error,
+            context: None,
+        }
+    }
+}
+
+pub struct CollectionParseError {
+    pub errors: Vec<ContextParseError>,
+    pub partial_collection: Option<Collection>,
+}
+
+impl From<ParseError> for CollectionParseError {
+    fn from(error: ParseError) -> Self {
+        Self {
+            errors: vec![error.into()],
+            partial_collection: None,
+        }
+    }
+}
+
+impl From<CollectionParseError> for ParseError {
+    fn from(mut error: CollectionParseError) -> Self {
+        error.errors.remove(0).error
+    }
+}
+
+pub fn collection_from_result(
+    result: Result<Collection, CollectionParseError>,
+) -> (Option<Collection>, Vec<ContextParseError>) {
+    match result {
+        Ok(collection) => (Some(collection), Vec::new()),
+        Err(err) => (err.partial_collection, err.errors),
+    }
+}
+
+fn display_context(context: &xmltree::Element) -> String {
+    let mut code = Vec::new();
+    let _ = context.write_with_config(
+        &mut code,
+        xmltree::EmitterConfig::new()
+            .write_document_declaration(false)
+            .perform_indent(true),
+    );
+    let code_string = String::from_utf8_lossy(&code).to_string();
+
+    match code_string.char_indices().nth(200) {
```

Comment:
Why the 200th exactly? i would write a single line comment about the reasoning if there is any

Reply:
Added a comment